### ISSUE TITLE
Descriptors and ClassVars

### ISSUE THREAD

**Issue by klen** [python/mypy#14969]
**Bug Report**

Have came across this [issue from pyright](https://github.com/microsoft/pyright/issues/4837)

**To Reproduce**

```python
# This sample tests the case where a member access is performed through
# an object using a field that is annotated as a ClassVar. Normally this
# is disallowed, but it is permitted if the type of the ClassVar is
# a descriptor object.
from __future__ import annotations

from typing import TYPE_CHECKING, ClassVar, Generic, Self, TypeVar, overload

T = TypeVar("T")


class Descriptor(Generic[T]):
    if TYPE_CHECKING:

        @overload
        def __get__(self, instance: None, owner) -> Self:
            ...

        @overload
        def __get__(self, instance: object, owner) -> T:
            ...

        def __get__(self, instance: object | None, owner) -> Self | T:
            ...

        def __set__(self, instance: object, value: T) -> None:
            ...

        def is_null(self) -> bool:
            ...


class Example:
    field1: ClassVar = Descriptor[str]()

    field2: ClassVar = ""

    def reset(self) -> None:
        self.field1 = ""

        # This should generate an error because field2 isn't
        # a descriptor object.
        self.field2 = ""

```

**Expected Behavior**

Copied @erictraut comment:

`ClassVar should work fine. If the class variable is assigned a descriptor, then the class variable will not be overwritten if you set the same attribute at the object level. So it sounds like mypy has a bug here. Pyright had a similar bug, which I recently fixed. Please report the bug to the maintainers of mypy.`

**Actual Behavior**

```
classvar.py:39: error: Cannot assign to class variable "field1" via instance  [misc]
classvar.py:43: error: Cannot assign to class variable "field2" via instance  [misc]
```

**Your Environment**

- Mypy version used: 1.1.1
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.11.2

**Comment by lhanekamp** [comment]
Hello! I would like to work on this issue can you please assign it to me?

**Comment by sobolevn** [comment]
@lhanekamp feel free to take this issue :)
We usually don't assign people on issues, it does not change anything.

If you have any questions, please feel free to ask them here!
